leakcanary-android-leak-deobfuscation-plugin/src/main/java/com/squareup/leakcanary/leak/deobfuscation/plugin/LeakcanaryLeakDeobfuscationPlugin.kt

```diff
@@ -0,0 +1,109 @@
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.ApplicationVariant
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.api.LibraryVariant
+import org.gradle.api.DefaultTask
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.UnknownTaskException
+import org.gradle.api.tasks.TaskProvider
+
+class LeakcanaryLeakDeobfuscationPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    val leakCanaryExtension =
+      project.extensions.create("leakCanary", LeakCanaryDeobfuscationExtension::class.java)
+
+    val androidAppPluginHandler = { _: Plugin<*> ->
+      val appExtension = project.extensions.getByType(AppExtension::class.java)
+      appExtension.applicationVariants.all { variant ->
+        setupPlugin(project, variant, leakCanaryExtension)
+      }
+    }
+
+    val androidLibraryPluginHandler = { _: Plugin<*> ->
+      val libraryExtension = project.extensions.getByType(LibraryExtension::class.java)
+      libraryExtension.libraryVariants.all { variant ->
+        setupPlugin(project, variant, leakCanaryExtension)
+      }
+    }
+
+    when {
+      project.plugins.hasPlugin("com.android.application") -> {
+        project.plugins.withId("com.android.application", androidAppPluginHandler)
+      }
+      project.plugins.hasPlugin("com.android.library") -> {
+        project.plugins.withId("com.android.library", androidLibraryPluginHandler)
+      }
+      else -> {
+        throw GradleException(
+            "This plugin can only be applied in Android application or library modules."
+        )
+      }
+    }
+  }
+
+  private fun setupPlugin(
+    project: Project,
+    variant: BaseVariant,
+    leakCanaryExtension: LeakCanaryDeobfuscationExtension
+  ) {
+    if (leakCanaryExtension.filterObfuscatedVariants(variant)) {
+      val copyObfuscationMappingFileTaskProvider = project.tasks.register(
+          "leakCanaryCopyObfuscationMappingFor${variant.name.capitalize()}",
+          CopyObfuscationMappingFileTask::class.java
+      ) {
+        it.variantDirName = variant.dirName
+        it.mappingFile = variant.mappingFile
+        it.mergeAssetsDirectory = variant.mergeAssetsProvider.get()
+            .outputDir.get()
+            .asFile
+      }
+
+      val mappingGeneratingTaskProvider =
+        findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithR8For${variant.name.capitalize()}"
+        ) ?: findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithProguardFor${variant.name.capitalize()}"
+        )
+
+      getPackageTaskProvider(variant)?.configure {
+        it.dependsOn(copyObfuscationMappingFileTaskProvider)
+      }
+      copyObfuscationMappingFileTaskProvider.configure { copyProguardMappingFileTask ->
```

Comment:
At this point if `copyObfuscationMappingFileTaskProvider` is null, there is a problem : devs passed a filter that filters in a variant that is not obfuscated. We should throw an exception on line 75.

Reply:
If we do that then it'll throw if android plugin hasn't been configured before. I tested it like this:
put deobfuscation plugin at  the top of build.gradle file in android-sample project
try to sync - it'll throw an error saying that the plugin can be applied only on android/library module
move deobfuscation plugin below android plugin and sync again

now during sync it'll throw on this line because it tries to find a task by name and this task doesn't exist at this point. I left it as is with null check, I talked to Xavier about this and he said  that currently there is no better way of finding those tasks. It should be better once AGP 3.7 is out.

Reply:
Yes, gradle plugins should be added in order. It would be better to start the plugin by this check and be sure we are in the right position for the rest of the plugin's code. What is the feature that will make it easier in 3.7 ?

Reply:
I think wasn't clear enough in previous message. If you have an android project that was successfully synced and you add leak canary  plugin then it works fine. But if you  put  leak canary plugin after android application one then sync will fail because LC plugin throws an exception (and it's good that it throws). The problem is that if you place LC plugin in correct place after an unsuccessful sync then next sync will also fail. You have to remove LC plugin, sync and then add it.
I'm not sure but I believe that it's because LC plugin is applied after android plugin but android plugin tasks were not created yet. I don't know if there is any way to wait until all tasks are created.

3.7 will introduce new variant api that will expose nice API to plug into the process of generating files like dex, or obfuscation mapping. It was clearly  explained during Android Dev Summit 2019, there is one talk about gradle.

What do you suggest to do now?

Reply:
@mzgreen I don't understand what you are explaining w.r.t to syncs and failed syncs. 
Gradle syncs are isolated, every sync restarts from scratch, there is no preservation of the state between syncs. 

Why would our plugin fail if we put it after the android plugin ? On what line ?

The way it works is that: devs will apply our plugin after they apply android plugin.
In our plugin we check the extension that is created immediatly by the android plugin.
Our plugin uses `variants.all`, which is asynchronous: when the android plugin will run, it will create variants, in the future, and we will react to it by adding our tasks or not.

Reply:
We have this:
```
      val mappingGeneratingTaskProvider =
        findTaskProviderOrNull(
            project,
            "transformClassesAndResourcesWithR8For${variant.name.capitalize()}"
        ) ?: findTaskProviderOrNull(
            project,
            "transformClassesAndResourcesWithProguardFor${variant.name.capitalize()}"
        )
```
If I throw an exception in case this is `null` then it fails during sync after an unsuccessful one. It tries to find this provider but it can't. Maybe I'm doing something else wrong and this code gets executed too early?

Reply:
I definitely don't understand this part: `it fails during sync after an unsuccessful one`. I am saying that after we try to find both tasks, we should have found one. If not, then we should crash. Again, there is no state preservation between syncs.

Reply:
It doesn't make sense to me either. 
I did this:
```
val mappingGeneratingTaskProvider =
        findTaskProviderOrNull(
            project,
            "transformClassesAndResourcesWithR8For${variant.name.capitalize()}"
        ) ?: findTaskProviderOrNull(
            project,
            "transformClassesAndResourcesWithProguardFor${variant.name.capitalize()}"
        ) ?: throw GradleException("Test")
```

Then in android sample project build.gradle file I moved LC plugin to the top:
```
apply plugin: 'com.squareup.leakcanary.deobfuscation'
apply plugin: 'com.android.application'
apply plugin: 'kotlin-android'
```

I ran sync and it failed and printed `Test` exception message (as expected).

Next I moved LC plugin to the bottom:
```
apply plugin: 'com.android.application'
apply plugin: 'kotlin-android'
apply plugin: 'com.squareup.leakcanary.deobfuscation'
```

And ran sync. This  time it should work but it still fails with  `Test` exception message.

I have to completely remove the plugin from sample app's build script, run sync and after it succeeds then I can add LC plugin back and run sync.  So it seems like something is being cached. Maybe AndroidStudio adds some caching here?

Reply:
@stephanenicolas
I think I've figured it out. Leak Canary sample project  uses this config:
```
    // Build with ./gradlew leakcanary-android-sample:installDebug -Pminify
    if (project.hasProperty('minify')) {
      debug {
        minifyEnabled true
        proguardFiles getDefaultProguardFile('proguard-android-optimize.txt')
      }
    } else {
      debug
    }
```

so if you don't pass `-Pminify` flag then minification is not enabled. It means that there is no variant with minification enabled in this case. When we sync the project in AS, there is no `-Pminify` flag so we don't have a variant with minification. It makes sense that it crashes now. But maybe we could give user a hint that he/she shouldn't be using this plugin when there is no variant with minification enabled? What do you think?

Reply:
yes, we could print something to users using logger.debug (not logger.lifecycle). Great you could find it. We can also print a message if we don't detect any android plugin saying that the plugin can only be applied to android project and it should be applied after the android plugin is applied.